### PR TITLE
RPED can now add parts to in-progress machine frames

### LINKED ISSUE #17352: PartExchangerSystem has a lot of dirty LINQ calls
plz kill

### PR DESCRIPTION
**Pull request by TotallyLemon**
## About the PR

This PR makes it so that the RPED can add parts to in-progress machine frames so that you don't have to manually put parts in; works just like exchanging in that it will pick the parts with the highest rating within the RPED. I decided to make it so that adding parts only takes 1 second compared to exchanging's 3 seconds in order to make it actually faster to use the RPED.

Based on this suggestion #16435
Fixes  #17352

**Media**

https://github.com/space-wizards/space-station-14/assets/79545393/e60d1116-734d-4efd-abc4-233f5ed80f2a

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Nanotrasen's R&D department have improved the RPED's design, allowing it to quickly install parts into in-progress machine frames.

### CODE REVIEW

**Comment by TotallyLemon** [comment]
Epic funny first PR moment

File: Content.Server/Construction/PartExchangerSystem.cs [hunk 1]
```diff
@@ -35,30 +35,83 @@ private void OnDoAfter(EntityUid uid, PartExchangerComponent component, DoAfterE
...
+
         var board = machine.BoardContainer.ContainedEntities.FirstOrNull();
 
         if (board == null || !TryComp<MachineBoardComponent>(board, out var macBoardComp))
             return;
 
-        var machineParts = new List<MachinePartComponent>();
-
-        foreach (var ent in storage.Storage.ContainedEntities) //get parts in RPED
+        foreach (var ent in new List<EntityUid>(machine.PartContainer.ContainedEntities)) //clone so don't modify during enumeration
```
Comment (metalgearsloth) [inline Content.Server/Construction/PartExchangerSystem.cs:65]:
```suggestion
        foreach (var ent in new ValueList<EntityUid>(machine.PartContainer.ContainedEntities)) //clone so don't modify during enumeration
```

At least if it takes a ctor.

File: Content.Server/Construction/PartExchangerSystem.cs [hunk 2]
```diff
@@ -35,30 +35,83 @@ private void OnDoAfter(EntityUid uid, PartExchangerComponent component, DoAfterE
...
         {
             if (TryComp<MachinePartComponent>(ent, out var part))
+            {
                 machineParts.Add(part);
+                _container.RemoveEntity(uid, ent);
+            }
         }
+
+        //order by highest rating
+        machineParts = machineParts.OrderByDescending(p => p.Rating).ToList();
```
Comment (metalgearsloth) [inline Content.Server/Construction/PartExchangerSystem.cs:75]: Use Sort, not LINQ.

File: Content.Server/Construction/PartExchangerSystem.cs [hunk 3]
```diff
@@ -35,30 +35,83 @@ private void OnDoAfter(EntityUid uid, PartExchangerComponent component, DoAfterE
...
+
+        var updatedParts = new List<MachinePartComponent>();
+        foreach (var (type, amount) in macBoardComp.Requirements)
+        {
+            var target = machineParts.Where(p => p.PartType == type).Take(amount);
```
Comment (metalgearsloth) [inline Content.Server/Construction/PartExchangerSystem.cs:80]: Same here, don't use LINQ.

File: Content.Server/Construction/PartExchangerSystem.cs [hunk 4]
```diff
@@ -83,9 +140,6 @@ private void OnDoAfter(EntityUid uid, PartExchangerComponent component, DoAfterE
             storage.Storage.Insert(unused.Owner);
             _storage.Insert(uid, unused.Owner, null, false);
         }
-        _construction.RefreshParts(args.Args.Target.Value, machine);
-
-        args.Handled = true;
```
Comment (metalgearsloth) [inline Content.Server/Construction/PartExchangerSystem.cs:88]: Why were these removed?

File: Content.Server/Construction/PartExchangerSystem.cs [hunk 5]
```diff
@@ -35,30 +35,83 @@ private void OnDoAfter(EntityUid uid, PartExchangerComponent component, DoAfterE
...
         if (!TryComp<ServerStorageComponent>(uid, out var storage) || storage.Storage == null)
             return; //the parts are stored in here
 
+        var machineParts = new List<MachinePartComponent>();
+
+        foreach (var ent in storage.Storage.ContainedEntities) //get parts in RPED
+        {
+            if (TryComp<MachinePartComponent>(ent, out var part))
+                machineParts.Add(part);
+        }
```
Comment (metalgearsloth) [inline Content.Server/Construction/PartExchangerSystem.cs:47]: Use GetEntityQuery<MachinePartComponent>

File: Content.Server/Construction/PartExchangerSystem.cs [hunk 6]
```diff
@@ -35,30 +35,83 @@ private void OnDoAfter(EntityUid uid, PartExchangerComponent component, DoAfterE
...
         if (board == null || !TryComp<MachineBoardComponent>(board, out var macBoardComp))
             return;
 
-        var machineParts = new List<MachinePartComponent>();
-
-        foreach (var ent in storage.Storage.ContainedEntities) //get parts in RPED
+        foreach (var ent in new List<EntityUid>(machine.PartContainer.ContainedEntities)) //clone so don't modify during enumeration
         {
             if (TryComp<MachinePartComponent>(ent, out var part))
+            {
```
Comment (metalgearsloth) [inline Content.Server/Construction/PartExchangerSystem.cs:68]: Same here on entityquery.

File: Content.Server/Construction/Components/PartExchangerComponent.cs [hunk 7]
```diff
@@ -12,6 +12,12 @@ public sealed class PartExchangerComponent : Component
     [DataField("exchangeDuration")]
     public float ExchangeDuration = 3;
 
+    /// <summary>
+    /// How long it takes to add the parts to an in-progress machine frame
+    /// </summary>
+    [DataField("constructDuration")]
+    public float ConstructDuration = 1;
```
Comment (metalgearsloth) [inline Content.Server/Construction/Components/PartExchangerComponent.cs:19]: This should just use exchangeduration.

**Review by metalgearsloth — DISMISSED** [review]

**Comment by metalgearsloth** [comment]
@EmoGarbage404

File: Content.Server/Construction/PartExchangerSystem.cs (hunk 4 above)
Comment (TotallyLemon) [reply Content.Server/Construction/PartExchangerSystem.cs:88]: Machine frames don't have a RefreshParts function and args are handled at the new line 52 in the OnDoAfter

**Comment by AJCM-git** [comment]
@TotallyLemon You coming back to this?

**Review by AJCM-git — CHANGES_REQUESTED** [review]
Doesn't this remove the ability of the RPED swap parts out of a finished machine?

**Comment by TotallyLemon** [comment]
> Doesn't this remove the ability of the RPED swap parts out of a finished machine?

You are still able to swap parts out of a finished machine.

File: Content.Server/Construction/PartExchangerSystem.cs [hunk 8]
```diff
@@ -35,35 +36,89 @@ private void OnDoAfter(EntityUid uid, PartExchangerComponent component, DoAfterE
...
+        TryExchangeMachineParts(args.Args.Target.Value, storage, machineParts);
+        TryConstructMachineParts(args.Args.Target.Value, storage, machineParts);
+
+        args.Handled = true;
+    }
+
+    private void TryExchangeMachineParts(EntityUid uid, ServerStorageComponent storage, List<MachinePartComponent> machineParts)
+    {
+        if (!TryComp<MachineComponent>(uid, out var machine) || storage.Storage == null)
+            return;
```
Comment (EmoGarbage404) [inline Content.Server/Construction/PartExchangerSystem.cs:60]: this would more typically be written by passing in the ServerStorageComponent as nullable with a default value of null, then calling `Resolve(uid, ref storage)`.

Comment (EmoGarbage404) [reply Content.Server/Construction/PartExchangerSystem.cs:60]: so:
```
    private void TryExchangeMachineParts(EntityUid uid, List<MachinePartComponent> machineParts,  ServerStorageComponent? storage = null, MachineComponent? machine = null)
    {
        if (!Resolve(uid, ref storage, ref machine))
            return;
```

File: Content.Server/Construction/PartExchangerSystem.cs [hunk 9]
```diff
@@ -83,9 +142,8 @@ private void OnDoAfter(EntityUid uid, PartExchangerComponent component, DoAfterE
             storage.Storage.Insert(unused.Owner);
             _storage.Insert(uid, unused.Owner, null, false);
         }
-        _construction.RefreshParts(args.Args.Target.Value, machine);
 
-        args.Handled = true;
```
Comment (EmoGarbage404) [inline Content.Server/Construction/PartExchangerSystem.cs:88]: pretty sure you still want this at the end

**Review by EmoGarbage404 — COMMENTED** [review]
apparently i underwent psychosis at the time of my initial comment.

File: Content.Server/Construction/PartExchangerSystem.cs (hunk 8 above)
Comment (TotallyLemon) [reply Content.Server/Construction/PartExchangerSystem.cs:60]: This causes a dereference of a possibly null reference error on line 94 but it would work if I did:
```
private void TryExchangeMachineParts(EntityUid uid, List<MachinePartComponent> machineParts, ServerStorageComponent? storage = null, MachineComponent? machine = null)
    {
        if (!Resolve(uid, ref storage, ref machine) || storage.Storage == null)
            return;
``` 

I just don't know if this is proper coding conventions.

Comment (TotallyLemon) [reply Content.Server/Construction/PartExchangerSystem.cs:60]: After testing it, I found that this code just causes the exchanging to not work at all. So I guess I'll keep it as is.

File: Content.Server/Construction/PartExchangerSystem.cs (hunk 9 above)
Comment (TotallyLemon) [reply Content.Server/Construction/PartExchangerSystem.cs:88]: It's at the end of the OnDoAfter method at line 54.
```
        TryExchangeMachineParts(args.Args.Target.Value, storage, machineParts);
        TryConstructMachineParts(args.Args.Target.Value, storage, machineParts);

        args.Handled = true;
    }
```
